firecrown/cosmosis/run.py

```diff
@@ -0,0 +1,220 @@
+    from cosmosis.runtime.utils import stdout_redirected
+    from cosmosis.runtime.config import Inifile
+except ImportError:
+    cosmosis = None
+
+# Locate the path to this directory
+# For fiddly reasons that make a great deal of sense in an entirely
+# different context to this one, we pass cosmosis a full path
+# to the interface file
+THIS_DIRECTORY = pathlib.Path(__file__).parent.resolve()
+COSMOSIS_INTERFACE = str(THIS_DIRECTORY.joinpath('interface.py'))
+
+
+def run_cosmosis(config, data):
+    """Run CosmoSIS on the problem.
+
+    This requires the following parameters 'sampler' section
+    of the config:
+
+      sampler - name of sampler to use, e.g. emcee, multinest, grid, ...
+      output - name of file to save to
+
+
+    Parameters
+    ----------
+    config : dict
+        Configuration info, usually read directly from the YAML file
+    data : dict
+        The result of calling `firecrown.config.parse` on an input YAML
+        config.
+    """
+
+    if cosmosis is None:
+        raise ImportError("CosmoSIS is not installed. "
+                          "See readme for instructions on doing so.")
+
+    # Extract the bits of the config file that
+    # cosmosis wants
+    ini = _make_cosmosis_params(config['cosmosis'])
+    values = _make_cosmosis_values(config['parameters'])
+    pool = _make_parallel_pool(config['cosmosis'])
+    pipeline = _make_cosmosis_pipeline(data, values, pool)
+
+    # Actually run the thing
+    cosmosis.main.run_cosmosis(None, pool=pool, ini=ini,
+                               pipeline=pipeline, values=values)
+
+    if pool is not None:
+        pool.close()
+
+
+def _make_parallel_pool(cosmosis_config):
+    """ Set up a parallel process pool.
+
+    Will look for the 'mpi' key in the cosmosis_config.
+
+    Parameters
+    ----------
+    cosmosis_config: dict
+        Sampler configuration section of the input
+
+
+    Returns
+    -------
+    pool: CosmoSIS MPIPool object
+        parallel process pool
+    """
+
+    # There is a reason to make the user actively
+    # request to use MPI rather than just checking -
+    # on many systems, including, importantly, NERSC,
+    # trying to import MPI when not running under the
+    # MPI environment will cause a crash
+    use_mpi = cosmosis_config.get('mpi', False)
+
+    if use_mpi:
+        pool = MPIPool()
+        if pool.size == 1:
+            print("Have mpi=True, but only running a single process.")
+            print("I will ignore and run in serial mode.")
+            pool = None
+    else:
+        pool = None
+        print("Running in serial mode (one process).")
+
+    return pool
+
+
+def _make_cosmosis_pipeline(data, values, pool):
+    """ Build a CosmoSIS pipeline.
+
+    Parameters
+    ----------
+    data: dict
+        The data object parse'd from an input yaml file.
+        This is passed as-is to the likelihood function
+
+    values: Inifile
+        Cosmosis object representing the input parameter values
+
```

Comment:
And this one too.

Reply:
you want to get rid of the lines between the parameters?  That makes it very hard to read.

Reply:
This is standard numpy doc conventions and CCL is using google which does similar things. shrug